src/main/java/org/commonjava/util/partyline/lock/global/impl/InfinispanGLM.java

```diff
@@ -0,0 +1,199 @@
+    {
+        this.lockCache = lockCache;
+        this.id = UUID.randomUUID().toString();
+    }
+
+    @Override
+    public boolean tryLock( String path, LockLevel level, long timeoutInMillis ) throws PartylineException
+    {
+        long cur = currentTimeMillis();
+        long end;
+        if ( timeoutInMillis <= 0 )
+        {
+            end = Long.MAX_VALUE;
+        }
+        else
+        {
+            end = cur + timeoutInMillis;
+        }
+
+        TransactionManager txManager = lockCache.getAdvancedCache().getTransactionManager();
+
+        while ( cur < end )
+        {
+            Boolean locked = executeInTransaction( txManager, () -> {
+                GlobalLockOwner lock = lockCache.get( path );
+                if ( lock == null )
+                {
+                    putWithExpiration( path, new GlobalLockOwner( level ).withOwner( this.id ) );
+                    return true;
+                }
+
+                // if some node is reading and this node want to read too, add this to owner list
+                if ( lock.getLevel() == LockLevel.read && level == LockLevel.read )
+                {
+                    if ( !lock.containsOwner( this.id ) )
+                    {
+                        lock.withOwner( this.id );
+                        putWithExpiration( path, lock );
+                    }
+                    return true;
+                }
+                return false;
+            } );
+
+            if ( locked == null || !locked )
+            {
+                sleepQuietly( 100 );
+            }
+            cur = currentTimeMillis();
+        }
+        return false;
+    }
+
+    @Override
+    public void unlock( String path, LockLevel level ) throws PartylineException
+    {
+        if ( level == LockLevel.write )
+        {
+            lockCache.remove( path );
+            return;
+        }
+
+        TransactionManager txManager = lockCache.getAdvancedCache().getTransactionManager();
+        Boolean ret = executeInTransaction( txManager, () -> {
+            GlobalLockOwner lock = lockCache.get( path );
+            if ( lock == null )
+            {
+                return true;
+            }
+            if ( lock.containsOwner( this.id ) )
+            {
+                lock.removeOwner( this.id );
+                if ( lock.isEmpty() )
+                {
+                    lockCache.remove( path );
+                }
+                else
+                {
+                    putWithExpiration( path, lock );
+                }
+            }
+            return true;
+        } );
+    }
+
+    private void putWithExpiration( String path, GlobalLockOwner lock )
+    {
+        lockCache.put( path, lock, DEFAULT_EXPIRATION_IN_MINUTES, TimeUnit.MINUTES );
+    }
+
+    private void sleepQuietly( long milliseconds )
+    {
+        try
+        {
+            sleep( milliseconds );
+        }
+        catch ( InterruptedException e )
+        {
+            ;
```

Comment:
Should we log this to TRACE at least? Just in case we need to see this happening? As long as any formatting happens in varargs / parameters, we can avoid the cost of formatting a string in case TRACE is turned off.

Reply:
Ok